File: src/features/typing/data/ts/merge_two_sorted_lists.ts
class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val = 0, next: ListNode | null = null) {
    this.val = val;
    this.next = next;
  }
}

function mergeTwoLists(list1: ListNode | null, list2: ListNode | null): ListNode | null {
  const dummy = new ListNode();
  let cur = dummy;
  while (list1 !== null && list2 !== null) {
    if (list1.val < list2.val) {
      cur.next = list1;
      list1 = list1.next;
    } else {
      cur.next = list2;
      list2 = list2.next;
    }
    cur = cur.next;
  }
  cur.next = list1 !== null ? list1 : list2;
  return dummy.next;
}

File: src/features/typing/data/dart/merge_two_sorted_lists.dart
class Solution {
  ListNode? mergeTwoLists(ListNode? list1, ListNode? list2) {
    final dummy = ListNode();
    var cur = dummy;
    var a = list1;
    var b = list2;
    while (a != null && b != null) {
      if (a.val < b.val) {
        cur.next = a;
        a = a.next;
      } else {
        cur.next = b;
        b = b.next;
      }
      cur = cur.next!;
    }
    cur.next = a ?? b;
    return dummy.next;
  }
}
